editor/src/components/canvas/controls/classname-select.tsx

```diff
@@ -196,31 +250,37 @@ export const ClassNameSelect: React.FunctionComponent = betterReactMemo('ClassNa
       return TailWindOptions.slice(0, MaxResults)
     } else {
       // First find all matches, and use a sparse array to keep the best matches at the front
-      let orderedMatchedResults: Array<Array<TailWindOption>> = []
-      let perfectMatchCount = 0
-      for (var i = 0; i < TailWindOptions.length && perfectMatchCount < MaxResults; i++) {
-        const nextOption = TailWindOptions[i]
-        const indexOf = nextOption.label.indexOf(trimmedLowerCaseInput)
-        if (indexOf > -1) {
-          let existingMatched = orderedMatchedResults[indexOf] ?? []
-          existingMatched.push(nextOption)
-          orderedMatchedResults[indexOf] = existingMatched
-          if (indexOf === 0) {
-            perfectMatchCount++
-          }
-        }
-      }
+      const orderedMatchedResults = findMatchingOptions(
+        trimmedLowerCaseInput,
+        TailWindOptions,
+        (option) => option.label,
+        MaxResults,
+      )
 
       // Now go through and take the first n best matches
-      let matchedResults: Array<TailWindOption> = []
-      let matchCount = 0
-
-      for (var j = 0; j < orderedMatchedResults.length && matchCount < MaxResults; j++) {
-        const nextMatches = orderedMatchedResults[j]
-        if (nextMatches != null) {
-          matchedResults.push(...nextMatches.slice(0, MaxResults - matchCount))
-          matchCount += nextMatches.length
-        }
+      let matchedResults = takeBestOptions(orderedMatchedResults, MaxResults)
+
+      // Next if we haven't hit our max result count, we find matches based on attributes
+      const remainingAllowedMatches = MaxResults - matchedResults.length
+      if (remainingAllowedMatches > 0) {
+        const orderedAttributeMatchedResults = findMatchingOptions(
+          trimmedLowerCaseInput,
+          AllAttributes,
+          (a) => a,
+          remainingAllowedMatches,
+        )
+        const bestMatchedAttributes = takeBestOptions(
+          orderedAttributeMatchedResults,
+          remainingAllowedMatches,
+        )
+        const classNamesFromBestMatchedAttributes = mapToClassNames(
+          bestMatchedAttributes,
+          remainingAllowedMatches,
+        )
+        const optionsForClassNames = classNamesFromBestMatchedAttributes.map((className) =>
+          TailWindOptions.find((option) => option.label === className),
```

Comment:
For each entry in `classNamesFromBestMatchedAttributes` this will (potentially) look through a ~40,000 entry array, that doesn't sound like it will be fast to me.

Reply:
Having tested this out, it doesn't noticeably impact the performance there, though I agree it does feel like there might be some cases that would.

I'm actually wondering how changes to the model will affect this though since we also want to show all attributes affected by a class in the menu, which means an upfront change to the options, which would in turn probably alter the searching.

I think I'm going to merge this as is so that it doesn't become stale, and then probably end up re-writing a lot of it in the very near term future anyway.